src/bp/core/mapping/mapping-repository.ts

```diff
@@ -0,0 +1,115 @@
+import { TYPES } from 'core/app/types'
+import Database from 'core/database'
+import { JobService } from 'core/distributed'
+import { inject, injectable, postConstruct } from 'inversify'
+
+import LRU from 'lru-cache'
+import ms from 'ms'
+
+@injectable()
+export class MappingRepository {
+  private repos: { [botId: string]: ScopedMappingRepository } = {}
+  private invalidateMappingCache: (scope: string, foreign: string, local: string) => void = this
+    ._localInvalidateMappingCache
+
+  constructor(
+    @inject(TYPES.Database) private database: Database,
+    @inject(TYPES.JobService) private jobService: JobService
+  ) {}
+
+  @postConstruct()
+  async init() {
+    this.invalidateMappingCache = <any>(
+      await this.jobService.broadcast<void>(this._localInvalidateMappingCache.bind(this))
+    )
+  }
+
+  forScope(scope: string) {
+    let repo = this.repos[scope]
+    if (!repo) {
+      repo = new ScopedMappingRepository(scope, this.database, (foreign, local) =>
+        this.invalidateMappingCache(scope, foreign, local)
+      )
+      this.repos[scope] = repo
+    }
+    return repo
+  }
+
+  private _localInvalidateMappingCache(scope: string, foreign: string, local: string) {
+    this.forScope(scope).localInvalidateCache(foreign, local)
+  }
+}
+
+class ScopedMappingRepository {
+  private readonly TABLE_NAME = 'mapping'
+
+  private localCache: LRU<string, string>
+  private foreignCache: LRU<string, string>
+
+  constructor(
+    private scope: string,
+    private database: Database,
+    private invalidateCache: (foreign: string, local: string) => void
+  ) {
+    this.localCache = new LRU({ max: 20000, maxAge: ms('5m') })
+    this.foreignCache = new LRU({ max: 20000, maxAge: ms('5m') })
+  }
+
+  public localInvalidateCache(foreign: string, local: string) {
+    this.localCache.del(foreign)
+    this.foreignCache.del(local)
+  }
+
+  async create(foreign: string, local: string) {
+    await this.query().insert({ scope: this.scope, foreign, local })
+
+    this.localCache.set(foreign, local)
+    this.foreignCache.set(local, foreign)
+  }
+
+  async delete(foreign: string, local: string) {
```

Comment:
I know the return types can be inferred, but can you still add them anyways?

Reply:
Yup